Src/Microsoft.Scripting/Runtime/SharedIO.cs

```diff
@@ -132,24 +127,21 @@ private sealed class StreamProxy : Stream {
             if (_outputStream == null) {
                 lock (_mutex) {
                     if (_outputStream == null) {
-#if FEATURE_FULL_CONSOLE
-    #if FEATURE_BASIC_CONSOLE
-                        if (!PreferBasicIO) {
-    #endif
-                            _outputStream = Console.OpenStandardOutput();
-                            _outputWriter = Console.Out;
-    #if FEATURE_BASIC_CONSOLE
-                            return;
+                        switch (ConsoleSupportLevel) {
+                            case SupportLevel.Full:
+                                _outputStream = Console.OpenStandardOutput();
+                                _outputWriter = Console.Out;
+                                break;
+                            case SupportLevel.Basic:
+                                _outputStream = new TextStream(Console.Out);
+                                _outputWriter = Console.Out;
+                                break;
+                            case SupportLevel.None:
+                            default:
+                                _outputStream = Stream.Null;
+                                _outputWriter = new StreamWriter(_outputStream, Encoding.UTF8, bufferSize: 16);
```

Comment:
The `StreamWriter` instead of `TextWriter.Null` is to have an encoding check?

Reply:
`TextWriter.Null` on .NET Framework uses a codepage encoding (probably `Encoding.Default`), so printing from ipy throws exceptions when the text to print contains characters that cannot be encoded in that codepage.

.NET, on the other hand, uses `Encoding.Unicode`, which can encode anything. I will modify the code so that it uses `StreamWriter` only when necessary.

Reply:
Hmm, interesting. Was assuming `TextWriter.Null` would just send stuff to the ether. Reference implementation (see below) itself (for .NET Framework) doesn't seem to make any use encoding so maybe we're the ones using it...
```cs
        [Serializable]
        private sealed class NullTextWriter : TextWriter
        {
            internal NullTextWriter(): base(CultureInfo.InvariantCulture) {
            }
        
            public override Encoding Encoding {
                get { return Encoding.Default;  }
            }
 
            [SuppressMessage("Microsoft.Contracts", "CC1055")]  // Skip extra error checking to avoid *potential* AppCompat problems.
            public override void Write(char[] buffer, int index, int count) {
            }
            
            public override void Write(String value) {
            }
 
            // Not strictly necessary, but for perf reasons
            public override void WriteLine() {
            }
 
            // Not strictly necessary, but for perf reasons
            public override void WriteLine(String value) {
            }
            
            public override void WriteLine(Object value) {
            }
        }
```

Reply:
Yes, `TextWriter.Null` swallows anything without problem. It is IronPython that complaints. I haven't investigated where specifically it happens, just observed the exception.

Reply:
> Is it possible that IronPython uses the encoding of the output writer to encode the text before sending it to the output stream directly?

Ah right, the `OutputEncoding` returns `_outputWriter.Encoding` and that's used to wrap the console stream in ipy.

Reply:
> Edit: kind of bad if you mutate your output after that's created.

Do you mean change `ConsoleSupportLevel` and call `RedirectToConsole()` after stdio files are created? In ipy, we can make sure it does not happen, and `SharedIO` access is [reserved to hosting providers like ipy](https://github.com/IronLanguages/dlr/blob/1767f9c5e6f1d168ebc9b91aa657e5af99e2588a/Src/Microsoft.Scripting/Hosting/Providers/HostingHelpers.cs#L18-L22).

Reply:
I guess I was thinking about the embedded scenario where people are having issues because they change the input/output  after their engine is created (whether done via `RedirectToConsole` or otherwise) and stdio encoding is locked down in the `TextIOWrapper`.

Anyway, not an issue with this - just something that popped to mind when I looked at the usage. Might be related to the recent email on the user group where someone called `_Engine.Runtime.IO.SetOutput(MyStream, Encoding.UTF8)` and print started outputting chinese characters.

Reply:
OK, thanks, it makes sense.